core/src/main/scala/org/apache/spark/util/ManualClock.scala

```diff
@@ -60,18 +66,21 @@ private[spark] class ManualClock(private var time: Long) extends Clock {
    */
   def waitTillTime(targetTime: Long): Long = synchronized {
     _isWaiting = true
+    _readyForFirstPeek = true
     try {
       while (time < targetTime) {
         wait(10)
       }
       getTimeMillis()
     } finally {
       _isWaiting = false
+      _readyForFirstPeek = false
     }
   }
 
   /**
-   * Returns whether there is any thread being blocked in `waitTillTime`.
+   * Returns whether there is any thread being blocked in `waitTillTime`, and this will be the first
+   * time it's been peeked in the blocking state.
    */
-  def isWaiting: Boolean = synchronized { _isWaiting }
+  def isWaitingAndReadyForFirstPeek: Boolean = synchronized { _isWaiting && _readyForFirstPeek }
```

Comment:
Guess it boils down to what is the purpose of _readyForPeek, which is not clear to me.